File: app/src/main/java/uwanttolearn/dagger2/java/app/di/AppComponent.java
package uwanttolearn.dagger2.java.app.di;

import android.app.Application;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjectionModule;
import uwanttolearn.dagger2.java.app.App;

/**
 * Created by waleed on 15/02/2018.
 */
@Component(modules = {
        AndroidInjectionModule.class,
        AppModule.class, ActivityProviderModule.class})
public interface AppComponent {

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder app(Application application);

        @BindsInstance
        Builder url(String url);


        AppComponent build();

    }

    void inject(App app);
}

File: app/src/main/java/uwanttolearn/dagger2/java/app/di/AppComponent.kt
package uwanttolearn.dagger2.java.app.di

import com.bumptech.glide.Glide
import dagger.Component
import uwanttolearn.dagger2.java.detail.DetailActivity
import uwanttolearn.dagger2.java.detail.di.DetailComponent
import uwanttolearn.dagger2.java.detail.di.DetailModule
import uwanttolearn.dagger2.java.home.di.HomeComponent
import uwanttolearn.dagger2.java.home.di.HomeModule
import uwanttolearn.dagger2.java.repositories.github.GitHubRepository

/**
 * Created by waleed on 30/12/2017.
 */
@Component(modules = arrayOf(AppModule::class))
interface AppComponent{

    fun plus(homeModule: HomeModule): HomeComponent

    fun plus(detailModule: DetailModule): DetailComponent



}